extensions/amp-a4a/0.1/test/test-amp-a4a.js

```diff
@@ -124,6 +139,198 @@ describe('amp-a4a', () => {
+        return a4a.adPromise_.then(() => {
+          // Force vsync system to run all queued tasks, so that DOM mutations
+          // are actually completed before testing.
+          a4a.vsync_.runScheduledTasks_();
+	  expect(getAdUrlSpy.calledOnce,
+		 'getAdUrlSpy called exactly once').to.be.true;
+        });
+      });
+    });
+
+    it('should throw error if getAdUrl not implemented', () => {
+      viewerWhenVisibleMock.onFirstCall().returns(Promise.resolve());
+      xhrMock.withArgs('https://test.location.org/ad/012345?args', {
+        mode: 'cors',
+        method: 'GET',
+        credentials: 'include',
+        requireAmpResponseSourceOrigin: true,
+      }).onFirstCall().returns(Promise.resolve(mockResponse));
+      return createAdTestingIframePromise().then(fixture => {
+        const doc = fixture.doc;
+        const a4aElement = createA4aElement(doc);
+        a4aElement.setAttribute('width', 200);
+        a4aElement.setAttribute('height', 50);
+        a4aElement.setAttribute('type', 'adsense');
+        const a4a = new MockA4AImplNoGetAdUrl(a4aElement);
+	const getAdUrlSpy = sandbox.spy(
+            a4a, 'getAdUrl');
+        doc.body.appendChild(a4aElement);
+        a4a.onLayoutMeasure();
+        expect(a4a.adPromise_).to.be.instanceof(Promise);
+        return a4a.adPromise_.then(() => {
+          // Force vsync system to run all queued tasks, so that DOM mutations
+          // are actually completed before testing.
+          a4a.vsync_.runScheduledTasks_();
+	  expect(getAdUrlSpy.calledOnce,
+		 'getAdUrlSpy called exactly once').to.be.true;
+	  expect(getAdUrlSpy.exceptions.length).to.equal(1);
+	});
+      });
+    });
+  });
+
+  describe('#sendXhrRequest_', () => {
+    it('should be called once', () => {
+      viewerWhenVisibleMock.onFirstCall().returns(Promise.resolve());
+      xhrMock.withArgs('https://test.location.org/ad/012345?args', {
+        mode: 'cors',
+        method: 'GET',
+        credentials: 'include',
+        requireAmpResponseSourceOrigin: true,
+      }).onFirstCall().returns(Promise.resolve(mockResponse));
+      return createAdTestingIframePromise().then(fixture => {
+        const doc = fixture.doc;
+        const a4aElement = createA4aElement(doc);
+        a4aElement.setAttribute('width', 200);
+        a4aElement.setAttribute('height', 50);
+        a4aElement.setAttribute('type', 'adsense');
+        const a4a = new MockA4AImpl(a4aElement);
+	const sendXhrRequest_Spy = sandbox.spy(
+            a4a, 'sendXhrRequest_');
+        doc.body.appendChild(a4aElement);
+        a4a.onLayoutMeasure();
+        expect(a4a.adPromise_).to.be.instanceof(Promise);
+        return a4a.adPromise_.then(() => {
+          // Force vsync system to run all queued tasks, so that DOM mutations
+          // are actually completed before testing.
+          a4a.vsync_.runScheduledTasks_();
+	  expect(sendXhrRequest_Spy.calledOnce,
+		 'sendXhrRequest_Spy called exactly once').to.be.true;
```

Comment:
Again, I think you'd like to check not just that the request is sent, but that the A4A code picks up / returns the response, as expected.